Plutarch/Api/V1/AssocMap.hs

```diff
@@ -47,3 +122,339 @@ instance
+    precList
+      ( \self x xs ->
+          pif
+            (pfstBuiltin # x #== key)
+            (apply #$ psndBuiltin # x)
+            (self # xs)
+      )
+      (const def)
+      # pto map
+
+-- | Insert a new key/value pair into the map, overiding the previous if any.
+pinsert :: (POrd k, PIsData k, PIsData v) => Term (s :: S) (k :--> v :--> PMap k v :--> PMap k v)
+pinsert = phoistAcyclic $
+  plam $ \key val ->
+    rebuildAtKey # (plam (pcons # (ppairDataBuiltin # pdata key # pdata val) #)) # key
+
+-- | Insert a new data-encoded key/value pair into the map, overiding the previous if any.
+pinsertData ::
+  (POrd k, PIsData k, PIsData v) =>
+  Term (s :: S) (PAsData k :--> PAsData v :--> PMap k v :--> PMap k v)
+pinsertData = phoistAcyclic $
+  plam $ \key val ->
+    rebuildAtKey # (plam (pcons # (ppairDataBuiltin # key # val) #)) # pfromData key
+
+-- | Delete a key from the map.
+pdelete :: (POrd k, PIsData k, PIsData v) => Term (s :: S) (k :--> PMap k v :--> PMap k v)
+pdelete = rebuildAtKey # plam id
+
+-- | Rebuild the map at the given key.
+rebuildAtKey ::
+  (POrd k, PIsData k) =>
+  Term
+    s
+    ( ( PBuiltinList (PBuiltinPair (PAsData k) (PAsData v))
+          :--> PBuiltinList (PBuiltinPair (PAsData k) (PAsData v))
+      )
+        :--> k
+        :--> PMap k v
+        :--> PMap k v
+    )
+rebuildAtKey = phoistAcyclic $
+  plam $ \handler key map ->
+    punsafeFrom $
+      precList
+        ( \self x xs ->
+            plet (pfromData $ pfstBuiltin # x) $ \k ->
+              plam $ \prefix ->
+                pif
+                  (k #< key)
+                  (self # xs #$ plam $ \suffix -> prefix #$ pcons # x # suffix)
+                  ( pif
+                      (k #== key)
+                      (prefix #$ handler # xs)
+                      (prefix #$ handler #$ pcons # x # xs)
+                  )
+        )
+        (const $ plam (#$ handler # pnil))
+        # pto map
+        # plam id
+
+-- | Construct an empty 'PMap'.
+pempty :: Term (s :: S) (PMap k v)
+pempty = punsafeFrom pnil
+
+-- | Construct a singleton 'PMap' with the given key and value.
+psingleton :: (PIsData k, PIsData v) => Term (s :: S) (k :--> v :--> PMap k v)
+psingleton = phoistAcyclic $ plam $ \key value -> psingletonData # pdata key # pdata value
+
+-- | Construct a singleton 'PMap' with the given data-encoded key and value.
+psingletonData :: (PIsData k, PIsData v) => Term (s :: S) (PAsData k :--> PAsData v :--> PMap k v)
+psingletonData = phoistAcyclic $
+  plam $ \key value -> punsafeFrom (pcons # (ppairDataBuiltin # key # value) # pnil)
+
+-- | Construct a 'PMap' from a list of key-value pairs, sorted by ascending key data.
+pfromAscList :: (POrd k, PIsData k, PIsData v) => Term (s :: S) (PBuiltinMap k v :--> PMap k v)
+pfromAscList = plam $ (passertSorted #) . pcon . PMap
+
+-- | Assert the map is properly sorted
+passertSorted :: (POrd k, PIsData k, PIsData v) => Term (s :: S) (PMap k v :--> PMap k v)
```

Comment:
? Shouldn't it always be sorted?

Reply:
This is meant for user-supplied values in datums and such.

Reply:
Well then it should be represented in the type, no? When would you ever have an unsorted map? Isn't this what `PTryFrom` is for?

Reply:
I didn't want to add a phantom type parameter to `AssocMap` as well, but that is an option. I need this function as a helper for `PValue.passertSorted`.

Reply:
Well, shouldn't all maps be sorted? You don't need to support unsorted maps.

Reply:
I've added a `KeyGuarantees = Sorted | Unsorted` type parameter to `PMap`, shared with `PValue`. This makes for a cleaner and safer interface overall. The downside is that `PLifted PMap` is now restricted to `Unsorted` because of the functional dependency.